File: sdk/graphql-transport/src/mappers/display.ts
// Copyright (c) Mysten Labs, Inc.
// Modifications Copyright (c) 2024 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import type { DisplayFieldsResponse } from '@iota/iota.js/client';

export function formatDisplay(object: {
    display?:
        | {
              key: string;
              value?: string | null | undefined;
              error?: string | null | undefined;
          }[]
        | null;
}) {
    const display: DisplayFieldsResponse = {
        data: null,
        error: null,
    };

    if (object.display) {
        object.display.forEach((displayItem) => {
            if (displayItem.error) {
                display!.error = displayItem.error as never;
            } else if (displayItem.value != null) {
                if (!display!.data) {
                    display!.data = {};
                }
                display!.data[displayItem.key] = displayItem.value;
            }
        });
    }

    return display;
}

File: crates/iota-types/src/display.rs
// Copyright (c) Mysten Labs, Inc.
// Modifications Copyright (c) 2024 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use serde::Deserialize;

use crate::{
    collection_types::VecMap,
    id::{ID, UID},
};

// TODO: add tests to keep in sync
/// Rust version of the Move iota::display::Display type
#[derive(Debug, Deserialize, Clone, Eq, PartialEq)]
pub struct DisplayObject {
    pub id: UID,
    pub fields: VecMap<String, String>,
    pub version: u16,
}

#[derive(Deserialize, Debug)]
/// The event that is emitted when a `Display` version is "released".
/// Serves for Display versioning.
pub struct DisplayVersionUpdatedEvent {
    pub id: ID,
    pub version: u16,
    pub fields: VecMap<String, String>,
}
